z3/src/solver.rs

```diff
@@ -205,9 +203,12 @@ impl<'ctx> Solver<'ctx> {
     /// - [`Config::set_proof_generation()`](struct.Config.html#method.set_proof_generation)
     ///
     /// [proof generation is not enabled]: struct.Config.html#method.set_proof_generation
-    pub fn get_proof(&self) -> Ast<'ctx> {
+    //
+    // TODO: what kind of AST does this actually return? If it's consistent, we can just
+    // make that the return type.
```

Comment:
In the Python bindings, this is labelled as an `ExprRef`, where the inheritance hierarchy is `AstRef` -> `ExprRef` -> (`BoolRef`, etc...).

But could we use an `-> impl Ast<'ctx>` here instead of `ast::Dynamic`?

Reply:
`-> impl Ast<'ctx>` seems to work in my testing.  In my simple example, the returned `Ast` had the sort `SortKind::Unknown`.

Reply:
Of course, we still have to use some concrete constructor in the method body.  In my testing, I was still calling `ast::Dynamic::new()` to construct the return value, so it's not terribly surprising I guess that the `-> impl Ast` worked as a declaration for the return type.

Reply:
In Python, they use `_to_expr_ref`:

```
def _to_expr_ref(a, ctx):
    if isinstance(a, Pattern):
        return PatternRef(a, ctx)
    ctx_ref = ctx.ref()
    k = Z3_get_ast_kind(ctx_ref, a)
    if k == Z3_QUANTIFIER_AST:
        return QuantifierRef(a, ctx)
    sk = Z3_get_sort_kind(ctx_ref, Z3_get_sort(ctx_ref, a))
    if sk == Z3_BOOL_SORT:
        return BoolRef(a, ctx)
    if sk == Z3_INT_SORT:
        if k == Z3_NUMERAL_AST:
            return IntNumRef(a, ctx)
        return ArithRef(a, ctx)
    if sk == Z3_REAL_SORT:
        if k == Z3_NUMERAL_AST:
            return RatNumRef(a, ctx)
        if _is_algebraic(ctx, a):
            return AlgebraicNumRef(a, ctx)
        return ArithRef(a, ctx)
    if sk == Z3_BV_SORT:
        if k == Z3_NUMERAL_AST:
            return BitVecNumRef(a, ctx)
        else:
            return BitVecRef(a, ctx)
    if sk == Z3_ARRAY_SORT:
        return ArrayRef(a, ctx)
    if sk == Z3_DATATYPE_SORT:
        return DatatypeRef(a, ctx)
    if sk == Z3_FLOATING_POINT_SORT:
        if k == Z3_APP_AST and _is_numeral(ctx, a):
            return FPNumRef(a, ctx)
        else:
            return FPRef(a, ctx)
    if sk == Z3_FINITE_DOMAIN_SORT:
        if k == Z3_NUMERAL_AST:
            return FiniteDomainNumRef(a, ctx)
        else:
            return FiniteDomainRef(a, ctx)
    if sk == Z3_ROUNDING_MODE_SORT:
        return FPRMRef(a, ctx)
    if sk == Z3_SEQ_SORT:
        return SeqRef(a, ctx)
    if sk == Z3_RE_SORT:
        return ReRef(a, ctx)
    return ExprRef(a, ctx)
```

But for the proof to be generated, the `check` / `check_assumptions` must have failed and proof generation must be enabled on the context.  In C, you'd do that via:

```
    Z3_set_param_value(cfg, "proof", "true");
```

Reply:
Right, which is called by `Config::set_proof_generation()` in Rust if I'm not mistaken.

So I guess this really is an `Ast` of no specialized type then.  We could create a minimal `Ast` subtype for this, we could even call it `Proof` if you want.  Is that the best way forwards?

Reply:
Let's leave it as is for now, I guess and then we can iterate on it and experiment and think.

Reply:
I interpret "as is" as "as `-> impl Ast`" for now, so that's in the last force-push. I think `-> impl Ast` is probably a better analogue to the Python `ExprRef` type than `Dynamic` is, anyway.